lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1993,7 +2036,188 @@ LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
+  // Identify read-write ports whose read end is unused.
+  llvm::SmallBitVector deadReads(op.getNumResults());
+  for (auto [i, port] : llvm::enumerate(op.getResults())) {
+    if (op.getPortKind(i) != MemOp::PortKind::ReadWrite)
+      continue;
+    if (!op.getPortAnnotation(i).empty())
+      continue;
+    if (isPortUnused(port, "rdata")) {
+      deadReads.set(i);
+      continue;
+    }
+  }
+  if (deadReads.none())
+    return false;
+
+  SmallVector<Type> resultTypes;
+  SmallVector<StringRef> portNames;
+  SmallVector<Attribute> portAnnotations;
+  for (auto [i, port] : llvm::enumerate(op.getResults())) {
+    if (deadReads[i])
+      resultTypes.push_back(
+          MemOp::getTypeForPort(op.getDepth(), op.getDataType(),
+                                MemOp::PortKind::Write, op.getMaskBits()));
+    else
+      resultTypes.push_back(port.getType());
+
+    portNames.push_back(op.getPortName(i));
+    portAnnotations.push_back(op.getPortAnnotation(i));
+  }
+
+  auto newOp = rewriter.create<MemOp>(
+      op.getLoc(), resultTypes, op.getReadLatency(), op.getWriteLatency(),
+      op.getDepth(), op.getRuw(), rewriter.getStrArrayAttr(portNames),
+      op.getName(), op.getNameKind(), op.getAnnotations(),
+      rewriter.getArrayAttr(portAnnotations), op.getInnerSymAttr(),
+      op.getGroupIDAttr());
+
+  for (unsigned i = 0, n = op.getNumResults(); i < n; ++i) {
+    auto result = op.getResult(i);
+    auto newResult = newOp.getResult(i);
+    if (deadReads[i]) {
+      // Create a wire to replace the old result. Wire the sub-fields of the
+      // old result to the relevant sub-fields of the write port.
+      auto wire = rewriter.create<WireOp>(result.getLoc(), result.getType());
+      result.replaceAllUsesWith(wire);
+
+      auto connect = [&](Value to, StringRef toName, Value from,
+                         StringRef fromName) {
+        auto toField = rewriter.create<SubfieldOp>(to.getLoc(), to, toName);
+        auto fromField =
+            rewriter.create<SubfieldOp>(from.getLoc(), from, fromName);
+        rewriter.create<StrictConnectOp>(result.getLoc(), toField, fromField);
+      };
+
+      connect(newResult, "addr", wire, "addr");
+      connect(newResult, "en", wire, "en");
+      connect(newResult, "clk", wire, "clk");
+      connect(newResult, "data", wire, "wdata");
+      connect(newResult, "mask", wire, "wmask");
+    } else {
+      result.replaceAllUsesWith(newResult);
+    }
+  }
+
+  rewriter.eraseOp(op);
+  return true;
+}
+
+LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
+  // Do not touch memories marked as such.
+  if (hasDontTouch(op.getOperation()))
+    return failure();
+
+  if (foldZeroWidthMemory(rewriter, op))
+    return success();
+  if (foldReadOrWriteOnlyMemory(rewriter, op))
+    return success();
+  if (foldReadWritePorts(rewriter, op))
+    return success();
+  if (foldUnusedPorts(rewriter, op))
+    return success();
```

Comment:
I was trying to think of a way to unify these functions, but couldn't. Can we write a single function, to iterate over the memory ports once, and just identify the appropriate pattern, instead of one function for each pattern, because each of these functions seems to follow a template. 
Maybe its not worth it.

Reply:
The methods do share some logic, but the transformations are different enough in my view to warrant different methods.